### PR TITLE
[Cart] Extract add to cart command and validate add simple product to cart

### PR DESCRIPTION
**Pull request by lchrusciel**

### CODE REVIEW

File: spec/Command/PutOptionBasedConfigurableItemToCartSpec.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    function it_has_quantity()
+    {
+        $this->quantity()->shouldReturn(5);
+    }
+
+    function it_throws_an_exception_if_order_token_is_not_a_string()
+    {
+        $this->beConstructedWith(new \StdClass(), 'T_SHIRT_CODE', ['RED_OPTION_CODE'], 1);
```
Comment (pamil) [inline spec/Command/PutOptionBasedConfigurableItemToCartSpec.php]: `\StdClass` -> `\stdClass`

File: spec/Command/PutOptionBasedConfigurableItemToCartSpec.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+use Sylius\ShopApiPlugin\Command\PutOptionBasedConfigurableItemToCart;
+
+final class PutOptionBasedConfigurableItemToCartSpec extends ObjectBehavior
+{
+    function let()
+    {
+        $this->beConstructedWith('ORDERTOKEN', 'T_SHIRT_CODE', ['RED_OPTION_CODE'], 5);
+    }
+
+    function it_is_initializable()
```
Comment (pamil) [inline spec/Command/PutOptionBasedConfigurableItemToCartSpec.php]: Just a reminder that this method is only a boilerplate to let phpspec work correctly :)

File: spec/Handler/PutOptionBasedConfigurableItemToCartHandlerSpec.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        $this->beConstructedWith($cartRepository, $productRepository, $cartItemFactory, $orderItemModifier, $orderProcessor, $manager);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType(PutOptionBasedConfigurableItemToCartHandler::class);
+    }
+
+    function it_handles_putting_new_item_to_cart(
+        OrderItemInterface $cartItem,
```
Comment (pamil) [inline spec/Handler/PutOptionBasedConfigurableItemToCartHandlerSpec.php]: What about using the constructor arguments first here?

File: spec/Handler/PutOptionBasedConfigurableItemToCartHandlerSpec.php [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+        $tShirt->getVariants()->willReturn([$blueTShirt, $redTShirt]);
+
+        $blueTShirt->getOptionValues()->willReturn([$blueOption]);
+        $blueOption->getCode()->willReturn('BLUE_OPTION_VALUE_CODE');
+        $blueOption->getOptionCode()->willReturn('COLOR_OPTION_CODE');
+
+        $redTShirt->getOptionValues()->willReturn([$redOption]);
+        $redOption->getCode()->willReturn('RED_OPTION_VALUE_CODE');
+        $redOption->getOptionCode()->willReturn('COLOR_OPTION_CODE');
```
Comment (pamil) [inline spec/Handler/PutOptionBasedConfigurableItemToCartHandlerSpec.php]: What's the difference between code and option code?

File: spec/Handler/PutOptionBasedConfigurableItemToCartHandlerSpec.php [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        $this->shouldHaveType(PutOptionBasedConfigurableItemToCartHandler::class);
+    }
+
+    function it_handles_putting_new_item_to_cart(
+        OrderItemInterface $cartItem,
+        OrderInterface $cart,
+        CartItemFactoryInterface $cartItemFactory,
+        OrderItemQuantityModifierInterface $orderItemModifier,
+        OrderProcessorInterface $orderProcessor,
+        OrderRepositoryInterface $cartRepository,
```
Comment (pamil) [inline spec/Handler/PutOptionBasedConfigurableItemToCartHandlerSpec.php]: `$orderRepository`?

File: spec/Validator/ProductExistsValidatorSpec.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType(ProductExistsValidator::class);
+    }
+
+    function it_is_constraint_validator()
+    {
+        $this->shouldHaveType( ConstraintValidator::class);
```
Comment (pamil) [inline spec/Validator/ProductExistsValidatorSpec.php]: CS 🗡

File: src/Controller/Cart/PutItemToCartAction.php [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+     *
+     * @return Response
+     */
+    public function __invoke(Request $request)
+    {
+        $commandRequest = $this->provideCommandRequest($request);
+
+        $validationResults = $this->validator->validate($commandRequest);
+
+        if (0 === count($validationResults)) {
```
Comment (pamil) [inline src/Controller/Cart/PutItemToCartAction.php]: What do you think about reverting this flow, e.g. early return for failing scenarios, happy path running till the end of method?

File: src/Request/PutOptionBasedConfigurableItemToCartRequest.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+     * @var int
+     */
+    private $quantity;
+
+    /**
+     * @param Request $request
+     */
+    public function __construct(Request $request)
+    {
+        $this->token = $request->attributes->get('token');
```
Comment (pamil) [inline src/Request/PutOptionBasedConfigurableItemToCartRequest.php]: Why don't we create this command here? We would get rid of these fields and have only one with command.

File: tests/Controller/CartPutItemToCartApiTest.php [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+    private static $acceptAndContentTypeHeader = ['CONTENT_TYPE' => 'application/json', 'ACCEPT' => 'application/json'];
+
+    /**
+     * @test
+     */
+    public function it_adds_a_product_to_the_cart()
+    {
+        $this->loadFixturesFromFile('shop.yml');
+
+        $token = 'SDAOSLEFNWU35H3QLI5325';
```
Comment (pamil) [inline tests/Controller/CartPutItemToCartApiTest.php]: What is it?

File: tests/Request/PutOptionBasedConfigurableItemToCartRequestTest.php [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        $putSimpleItemToCartRequest = new PutOptionBasedConfigurableItemToCartRequest(new Request([], [
+            'productCode' => 'HACKTOBERFEST_TSHIRT_CODE',
+            'options' => ['LARGE__CODE'],
+            'quantity' => 4,
+        ], ['token' => 'ORDERTOKEN']));
+
+        $command = $putSimpleItemToCartRequest->getCommand();
+
+        $this->assertInstanceOf(PutOptionBasedConfigurableItemToCart::class, $command);
```
Comment (pamil) [inline tests/Request/PutOptionBasedConfigurableItemToCartRequestTest.php]: `Assert::assertSame($command, new PutOptionBasedConfigurableItemToCart(...));`?

File: src/Request/PutOptionBasedConfigurableItemToCartRequest.php (hunk 8 above)
Comment (lchrusciel) [inline src/Request/PutOptionBasedConfigurableItemToCartRequest.php]: But there are 2 more requests already defined. Maybe I will do it in separate PR

Comment (lchrusciel) [inline src/Request/PutOptionBasedConfigurableItemToCartRequest.php]: That was a stupid answer. The reason is that we need these fields for validation purposes.

File: tests/Controller/CartPutItemToCartApiTest.php (hunk 9 above)
Comment (lchrusciel) [inline tests/Controller/CartPutItemToCartApiTest.php]: random token which is used for cart handling

**Pull request merged**